### PR TITLE
Polly retry for Database and Az Function call

### PR DESCRIPTION
**Pull request by kcherukupalli**
## What’s changing?
Added Polly retry policy for Database and Az Function calls.  Policy includes the following list of codes and will retry every 5 seconds for 3 times.
**For Database calls:**
      08000	connection_exception
      08003	connection_does_not_exist
      08006	connection_failure
      08001	sqlclient_unable_to_establish_sqlconnection
      08004	sqlserver_rejected_establishment_of_sqlconnection
      08007	transaction_resolution_unknown
      53300	too_many_connections

**For Az Function calls**
                HttpStatusCode.RequestTimeout,
                HttpStatusCode.TooManyRequests,
                HttpStatusCode.InternalServerError,
                HttpStatusCode.BadGateway,
                HttpStatusCode.ServiceUnavailable,
                HttpStatusCode.GatewayTimeout

## Why?
nac-1838

## This PR has:

- [x] **JIRA ticket(s)** (_Must be listed in PR request and branch name should follow convention NAC-123-description_)<description>

- [ ] **Security Related Changes** (_Security Tab questions have been answered on JIRA ticket(s)_)

- [x] **Commented source code** (_required on public classes and methods, and elsewhere as appropriate_)

- [ ] **Documented API changes** (_required on public classes and methods, and elsewhere as appropriate_)
    - [ ] External API documentation - OpenApi docs and Generated Docs (e.g. https://github.com/18F/piipan/tree/dev/docs/openapi) 
    - [ ] Internal API documentation (e.g. https://github.com/18F/piipan/tree/dev/metrics/docs/openapi)   

- [ ] **Developer documentation** (_for new build/test/API changes or complex portions of the system_)

- [x] **Automated unit tests** (_to maintain or increase level of code coverage_)

- [ ] **Changes to IAC** (_add any deployment steps that require manual intervention to the draft release notes_)
    - [ ] Have IaC scripts run successfully against an existing deployed environment?
    - [ ] Have IaC scripts run successfully against a newly deployed environment?
    - [ ] Have any new bash scripts been given the appropriate permissions? (e.g. git update-index --chmod=+x $script)
    - [ ] Do the IaC changes require deleting existing resources?

## Anything else?

List of codes needs to be verified.

## Code Review Checklist:

- [ ] Automated checks all pass? (code builds successfully, tests pass, test coverage meets/exceeds threshold, etc.) 
- [ ] Have all requirements and Acceptance Criteria been met?
- **If there are code changes**
    - [ ] Is there no unnecessary complexity? Code is readable, maintainable, etc.
    - [ ] Do the changes follow SOLID principles?
    - [ ] Are the changes robust? Concurrency considered? Performance? Security (e.g. no sql injections)?
    - [ ] Are newly added dependencies needed and appropriate? Are dependency files (e.g. packages.json & packages.lock.json) updated?
    - [ ] For any new projects/solutions, have they been added to code coverage files?
    - [ ] Public classes & methods are documented    
- **If there are API changes**
    - [ ] Does API make sense? Useful and concise?
    - [ ] Are existing API interfaces/contracts intact? i.e. no breaking changes
    - [ ] Are API tests included?
    - [ ] Are API responses appropriate and documented? (200s, 201, 404s, etc.)
    - [ ] Are internals kept private and not leaked into the API?
    - [ ] External API documentation - OpenApi docs and Generated Docs 
    - [ ] Internal API documentation    
- **If there are IAC changes**
    - [ ] IAC Scripts are readable & maintainable
    - [ ] IAC Scripts are documented
    - [ ] IAC scripts pass required checks
    - [ ] IAC scripts were run successfully by author
- **General Logging Questions**
    - [ ] Do new capabilities (or changed capabilities) include logging?
    - [ ] Is PII not logged?
- **General Error Handling Questions**
    - [ ] Has exception handling been added for new capabilities (or changed capabilities)?
    - [ ] Are errors/exceptions properly communicated to end-users? Were there requirements for user facing error messages?
    - [ ] Are errors/exceptions logged?
    - [ ] Is PII kept out of error/exception logging?
- **If there are Developer Documentation Changes**
    - [ ] General Project documentation (under piipan\docs. e.g. iac.md, architecture.md, etc.)
    - [ ] Github documentation additions/changes (Readmes)
- **General automated test questions**
    - [ ] End-to-End Tests provided (if code implements new user story)
    - [ ] Are tests sufficient?
    - [ ] Corner cases considered?
    - [ ] Are performance tests needed?

### CODE REVIEW

File: shared/src/Piipan.Shared/Database/BaseDatabaseManager.cs [hunk 1]
```diff
@@ -82,9 +84,18 @@ public abstract class BaseDatabaseManager<DbType> : IDatabaseManager<DbType>
         /// <returns></returns>
         public virtual async Task<T> PerformQuery<T>(Func<NpgsqlConnection, Task<T>> query, string database = null, CancellationToken cancellationToken = default)
         {
-            // TODO: Perform retries, throttle, log. All queries should flow through here.
-            await using var connection = await GetConnectionAsync(database?.ToLower(), cancellationToken);
-            return await query(connection);
+            var retryTimes = 3;
+            var retryableErrorCodes = new[] { "08000", "08001", "08007", "08006", "08004", "08003", "53300" };
```
Comment (bgfoshay) [inline shared/src/Piipan.Shared/Database/BaseDatabaseManager.cs:88]: Nitpick/Suggestion: It may be helpful here to list the corresponding exceptions for each code OR provide a link to page with that information.

File: shared/src/Piipan.Shared/Http/AuthorizedJsonApiClient.cs [hunk 2]
```diff
@@ -41,47 +43,81 @@ private async Task<HttpRequestMessage> PrepareRequest(string path, HttpMethod me
...
-        public async Task<TResponse> PostAsync<TRequest, TResponse>(string path, TRequest body)
+            var response = await Policy.HandleResult<HttpResponseMessage>(r => retryResponseCodes.Contains(r.StatusCode))
+                    .WaitAndRetryAsync(3, retryAttempt => TimeSpan.FromSeconds(5))
+                    .ExecuteAsync(async () => await BaseSendAsync(await PrepareRequest(path, HttpMethod.Get, headerFactory)));
+            return response;
+        }
+        private async Task<HttpResponseMessage> PrepareAndSendAsyncForPostAndPatch<TRequest>(string path, TRequest body, Func<IEnumerable<(string, string)>> headerFactory, HttpMethod httpMethod)
         {
-            return await PostAsync<TRequest, TResponse>(path, body, () => Enumerable.Empty<(string, string)>());
+            var retryResponseCodes = new HttpStatusCode[]
```
Comment (bgfoshay) [inline shared/src/Piipan.Shared/Http/AuthorizedJsonApiClient.cs:70]: Nitpick/Suggestion: We could probably pull these response codes up into a member variable and avoid duplicating.

File: shared/src/Piipan.Shared/Http/AuthorizedJsonApiClient.cs [hunk 3]
```diff
@@ -41,47 +43,81 @@ private async Task<HttpRequestMessage> PrepareRequest(string path, HttpMethod me
...
+            {
+                HttpStatusCode.RequestTimeout,
+                HttpStatusCode.TooManyRequests,
+                HttpStatusCode.InternalServerError,
+                HttpStatusCode.BadGateway,
+                HttpStatusCode.ServiceUnavailable,
+                HttpStatusCode.GatewayTimeout
+            };
+            var response = await Policy.HandleResult<HttpResponseMessage>(r => retryResponseCodes.Contains(r.StatusCode))
+                    .WaitAndRetryAsync(3, retryAttempt => TimeSpan.FromSeconds(5))
```
Comment (bgfoshay) [inline shared/src/Piipan.Shared/Http/AuthorizedJsonApiClient.cs:80]: Question: Will this retry policy compound on database connection failures? e.g. database connection failures cause 3 retries to connect the database. Once the failure bubbles up to this level, it'll receive the failure and kick into retry gear as well, resulting in a total of 9 retries?

File: shared/tests/Piipan.Shared.Tests/Database/SingleDatabaseManagerTests.cs [hunk 4]
```diff
@@ -234,5 +245,42 @@ public async Task PerformQuery_ManyTimes_SimultaneouslyWorksFine()
...
+
+        [Theory]
+        public async Task PerformQuery_For_ErrorCodesInPolicyForRetry(string errorCode)
+        {
+            // Arrange
+            var mockTokenProvider = DefaultMocks.MockAzureServiceTokenProvider();
+            var databaseManager = new SingleDatabaseManager<MockType>(
+                    ConnectionString, mockTokenProvider.Object);
+
+            // Act / Assert.   MultipleDatabaseManager  throw an error which is part of the Polly retry policy.
```
Comment (bgfoshay) [inline shared/tests/Piipan.Shared.Tests/Database/SingleDatabaseManagerTests.cs:263]: Nitpick: Should this comment be in regards to SingleDatabaseManager?

File: shared/tests/Piipan.Shared.Tests/Http/AuthorizedJsonApiClientTests.cs [hunk 5]
```diff
@@ -334,5 +337,189 @@ public async Task TryGetAsync_ThrowsErrorOnOtherStatus()
...
+        }
+
+        [InlineData(HttpStatusCode.RequestTimeout)]
+        [InlineData(HttpStatusCode.TooManyRequests)]
+        [InlineData(HttpStatusCode.InternalServerError)]
+        [InlineData(HttpStatusCode.BadGateway)]
+        [InlineData(HttpStatusCode.ServiceUnavailable)]
+        [InlineData(HttpStatusCode.GatewayTimeout)]
+        [Theory]
+        public async Task SendAsyncGet_Retries3Times_AsExpected(HttpStatusCode statusCode)
```
Comment (bgfoshay) [inline shared/tests/Piipan.Shared.Tests/Http/AuthorizedJsonApiClientTests.cs:359]: Question: 
1) Should we consider tests for codes NOT in policy retry?
2) Should we consider a test to make sure retries are not compounding? This relates to my above comment. Not sure if we need to implement a circuit breaker or something else to guard against this.

File: shared/src/Piipan.Shared/Database/BaseDatabaseManager.cs [hunk 6]
```diff
@@ -82,9 +84,18 @@ public abstract class BaseDatabaseManager<DbType> : IDatabaseManager<DbType>
...
+            AsyncRetryPolicy retryPolicy = Policy
+               .Handle<NpgsqlException>(ex => retryableErrorCodes.Any(errorCode => ex.Message.Contains(errorCode)))
```
Comment (dborgen2) [inline shared/src/Piipan.Shared/Database/BaseDatabaseManager.cs:90]: Ryan mentioned and I also remember seeing awhile back `PostgresException` exceptions being thrown. I think we should handle them as well (error codes should be the same).

On another note, should we be looking at ex.SqlState, instead of assuming the message contains the error code?

File: shared/src/Piipan.Shared/Http/AuthorizedJsonApiClient.cs [hunk 7]
```diff
@@ -41,47 +43,81 @@ private async Task<HttpRequestMessage> PrepareRequest(string path, HttpMethod me
...
+                    .ExecuteAsync(async () => await BaseSendAsync(await PrepareRequestForPostAndPatch(path, body, headerFactory, httpMethod)));
+            return response;
+
         }
 
-        public async Task<TResponse> PostAsync<TRequest, TResponse>(string path, TRequest body, Func<IEnumerable<(string, string)>> headerFactory)
+        private async Task<HttpRequestMessage> PrepareRequestForPostAndPatch<TRequest>(string path, TRequest body, Func<IEnumerable<(string, string)>> headerFactory, HttpMethod httpMethod)
```
Comment (dborgen2) [inline shared/src/Piipan.Shared/Http/AuthorizedJsonApiClient.cs:86]: Minor nitpick (non-blocking): Maybe we should call this PrepareRequestWithBody. We may want to add Put at some point and Put should be able to utilize this method too.

File: shared/tests/Piipan.Shared.Tests/Database/MultipleDatabaseManagerTests.cs [hunk 8]
```diff
@@ -31,6 +38,16 @@ private async Task<int> PerformDummyQuery(MultipleDatabaseManager<MockType> data
         {
             return await databaseManager.PerformQuery(conn => Task.FromResult(1), database);
         }
+        private async Task<int> PerformDummyQueryForRetry(MultipleDatabaseManager<MockType> databaseManager, string database = "ea", string errorCode="")
+        {
+            int count = 0;
+            
+            return await databaseManager.PerformQuery(conn => { 
+                count++;
+                throw new Npgsql.NpgsqlException(String.Format("For errorCode: {0}, No of retries: {1}.", errorCode, count.ToString()));
```
Comment (dborgen2) [inline shared/tests/Piipan.Shared.Tests/Database/MultipleDatabaseManagerTests.cs:47]: Minor nitpick (non-blocking): This message is a bit confusing. I think it should be No of tries. If for example it only performs this method once, we tried one time and retried zero times.

File: shared/tests/Piipan.Shared.Tests/Database/SingleDatabaseManagerTests.cs [hunk 9]
```diff
@@ -30,7 +32,16 @@ private async Task<int> PerformDummyQuery(SingleDatabaseManager<MockType> databa
         {
             return await databaseManager.PerformQuery(conn => Task.FromResult(1));
         }
+        private async Task<int> PerformDummyQueryForRetry(SingleDatabaseManager<MockType> databaseManager, string database = "ea", string errorCode = "")
+        {
+            int count = 0;
 
+            return await databaseManager.PerformQuery(conn => {
+                count++;
+                throw new Npgsql.NpgsqlException(String.Format("For errorCode: {0}, No of retries: {1}.", errorCode, count.ToString()));
```
Comment (dborgen2) [inline shared/tests/Piipan.Shared.Tests/Database/SingleDatabaseManagerTests.cs:41]: Minor nitpick (non-blocking): Same thought here. Should this same No of tries instead of No of retries?

File: shared/src/Piipan.Shared/Http/AuthorizedJsonApiClient.cs [hunk 10]
```diff
@@ -41,47 +43,81 @@ private async Task<HttpRequestMessage> PrepareRequest(string path, HttpMethod me
...
+            return requestMessage;
+        }
+        /// <summary>
+        /// Do not call this function directly. This method is split out for testing purposes.
+        /// </summary>
+        protected virtual Task<HttpResponseMessage> BaseSendAsync(HttpRequestMessage request)
+        {
+            return Client().SendAsync(request);
+        }
+        public async Task<TResponse> PostAsync<TRequest, TResponse>(string path, TRequest body)
```
Comment (dborgen2) [inline shared/src/Piipan.Shared/Http/AuthorizedJsonApiClient.cs:105]: Question (non-blocking): These methods were originally not commented, but since there were changes to most of these public methods (or at least Github says there is), should we comment them in case we would get flagged for it?

**Comment by dborgen2** [comment]
I'm having issues with Jira at the moment, but I only see one Jira ticket referenced here and in the branch name. I assume there is actually two (one for database and one for Api retries)?

File: shared/src/Piipan.Shared/Http/AuthorizedJsonApiClient.cs (hunk 3 above)
Comment (dborgen2) [reply shared/src/Piipan.Shared/Http/AuthorizedJsonApiClient.cs:80]: FYI, I can include this end-end test when I check in my PR for improving NAC-2126. But if there are any smaller tests Krishna you could check in the meantime, ie. making sure it doesn't throw a 500 error when the database retries, that would be great.

File: shared/src/Piipan.Shared/Database/BaseDatabaseManager.cs (hunk 6 above)
Comment (kcherukupalli) [reply shared/src/Piipan.Shared/Database/BaseDatabaseManager.cs:90]: Added PostgresException to the Policy and looking at  SqlState for retry.   Refactored unit test covering only PostgresException.  Took out the unit test covering NpgsqlException as SqlState is not exposed to mock the exception.

File: shared/tests/Piipan.Shared.Tests/Http/AuthorizedJsonApiClientTests.cs (hunk 5 above)
Comment (kcherukupalli) [reply shared/tests/Piipan.Shared.Tests/Http/AuthorizedJsonApiClientTests.cs:359]: Added condition for Not in policy for retry test.  End to end test case will be handled in a different PR.

File: shared/src/Piipan.Shared/Http/AuthorizedJsonApiClient.cs [hunk 11]
```diff
@@ -157,7 +172,12 @@ public async Task<TResponse> GetAsync<TResponse>(string path, string query = nul
 
             return JsonConvert.DeserializeObject<TResponse>(responseContentJson);
         }
-        
+        /// <summary>
+        /// GET  the response for an ASync call for the Request
```
Comment (dborgen2) [inline shared/src/Piipan.Shared/Http/AuthorizedJsonApiClient.cs:176]: Minor (nonblocking) - We should say something that if the response comes back with a 404 we do not throw an exception (which is the main difference between this and GetAsync)

File: shared/src/Piipan.Shared/Database/BaseDatabaseManager.cs [hunk 12]
```diff
@@ -32,6 +34,9 @@ public abstract class BaseDatabaseManager<DbType> : IDatabaseManager<DbType>
         protected string ConnectionString { get; }
         protected bool ShouldFetchAzureToken { get; }
 
+        public int NoOfRetries = 3;
```
Comment (dborgen2) [inline shared/src/Piipan.Shared/Database/BaseDatabaseManager.cs:37]: Minor (non-blocking) - This probably should be a property if it needs to be public.

**Review by dborgen2 — APPROVED** [review]
👍 LGTM, thanks for addressing the concerns. Just a few minor non-blocking things
